### DISCUSSION TITLE
Can't customize output path of Prisma client: Error: Dynamic require of "tty" is not supported

### DISCUSSION THREAD

**Discussion by coindegen** [prisma/prisma#22121] [Q&A] [answered]
### Question

Hello, I am attempting to customize the location of my Prisma client (https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#the-location-of-prisma-client) by adding an output field to my schema.prisma like this:

```
generator client {
  provider = "prisma-client-js"
  output   = "./client"
}
```

Instead of importing my Prisma client from the default node_modules (`import { PrismaClient } from '@prisma/client'`), I am importing from this custom location (`import { PrismaClient } from '../../prisma/client/index.js'`).

While my build was working fine in the former default location, it is breaking my build in the new location with this error message:

```
$ npm run dev

> predev
> npm run build:icons --silent


> dev
> remix dev -c "node ./server/dev-server.js" --manual


 💿  remix dev

 info  building...
 info  built (1.6s)
8:33:33 p.m. [tsx] rerunning
Debugger listening on ws://127.0.0.1:9229/bff68bdb-442c-4317-a22b-c77f1eb1ebaf
For help, see: https://nodejs.org/en/docs/inspector
🔶 Mock server installed
Error: Dynamic require of "tty" is not supported
Error: Dynamic require of "tty" is not supported
    at file:///Users/madhava/Sites/projects/epic-stack/build/index.js:11:9
    at file:///Users/madhava/Sites/projects/epic-stack/prisma/client/runtime/library.js:1:8469
    at file:///Users/madhava/Sites/projects/epic-stack/prisma/client/runtime/library.js:1:222
    at file:///Users/madhava/Sites/projects/epic-stack/prisma/client/runtime/library.js:5:272
    at file:///Users/madhava/Sites/projects/epic-stack/prisma/client/runtime/library.js:1:222
    at prisma/client/runtime/library.js (file:///Users/madhava/Sites/projects/epic-stack/prisma/client/runtime/library.js:7:7703)
    at file:///Users/madhava/Sites/projects/epic-stack/build/index.js:17:50
    at prisma/client/index.js (file:///Users/madhava/Sites/projects/epic-stack/prisma/client/index.js:24:5)
    at file:///Users/madhava/Sites/projects/epic-stack/build/index.js:17:50
    at app/utils/db.server.ts (file:///Users/madhava/Sites/projects/epic-stack/app/utils/db.server.ts:4:30)
```

A diff between the two client/index.js files shows they're pretty much identical:
https://gist.github.com/coindegen/fd33956bade50bcaf1c9f6f743002eb1

### How to reproduce (optional)

This error can be reproduced by the following:

1. Clone Kent C. Dodd's epic-stack: https://github.com/epicweb-dev/epic-stack
2. `npm install; cp .env.example .env`
3. `npm run dev` to confirm it's working
4. Make these small modifications: https://github.com/coindegen/epic-stack/pull/1/files
5. Run `npx prisma generate; npm run dev`

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: MacOS
- Database: SQLite
- Node.js version: 18

- Run `prisma -v` to see your Prisma version and paste it

```
prisma                  : 5.6.0
@prisma/client          : 5.6.0
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v18.18.2
Query Engine (Node-API) : libquery-engine e95e739751f42d8ca026f6b910f5a2dc5adeaeee (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli e95e739751f42d8ca026f6b910f5a2dc5adeaeee (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.6.0-32.e95e739751f42d8ca026f6b910f5a2dc5adeaeee
Default Engines Hash    : e95e739751f42d8ca026f6b910f5a2dc5adeaeee
Studio                  : 0.495.0
```

**Comment by ludralph** [comment]
Hi @coindegen 👋 

Thank you for raising this question.

I am able to reproduce this error. I will need some time to figure this out. I'll get back to you with updates I may have.

**Reply by coindegen** [reply]
Hi @ludralph , appreciate you looking into it! 🙏

I believe it is related to this issue, it came up in Prisma 3 as well: https://github.com/prisma/prisma/issues/10058#issuecomment-964117671

**Reply by coindegen** [reply]
tagging @millsp since he authored the PR that fixed this a few years ago: https://github.com/prisma/prisma/pull/10222

Is this a regression of the same bundling issue?

**Reply by coindegen** [reply]
In contrast, I didn't have this problem when I made the same change on another repo using the Prisma client:
https://github.com/coindegen/turso-prisma/pull/1/files

This (much simpler) repo uses "node" for moduleResolution rather than "nodenext", and there's no build step, I just run my migrate.ts script with `ts-node`.

**Reply by coindegen** [reply, marked as answer]
The workaround identified by @andyjy here works for me: https://github.com/prisma/prisma/issues/9435#issuecomment-974576800

By moving my output to node_modules like this:

```
 generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/@splinter/prisma/client"
}
```

And my import to this: `import { PrismaClient } from "@splinter/prisma/client/index.js";`

The error goes away.
